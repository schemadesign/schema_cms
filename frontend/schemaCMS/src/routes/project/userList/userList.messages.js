/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

export default defineMessages({
  headerTitle: {
    id: 'project.userList.headerTitle',
    defaultMessage: 'Users',
  },
  headerSubtitle: {
    id: 'project.userList.headerSubtitle',
    defaultMessage: 'Overview',
  },
  removeTitle: {
    id: 'project.userList.removeTitle',
    defaultMessage: 'Are you sure you want to remove the user from the project?',
  },
  cancelRemoval: {
    id: 'project.userList.cancelRemoval',
    defaultMessage: 'Cancel',
  },
  confirmRemoval: {
    id: 'project.userList.confirmRemoval',
    defaultMessage: 'Confirm',
  },
  back: {
    id: 'project.userList.back',
    defaultMessage: 'Back',
  },
  projectDetails: {
    id: 'project.userList.projectDetails',
    defaultMessage: 'Project details',
  },
});
