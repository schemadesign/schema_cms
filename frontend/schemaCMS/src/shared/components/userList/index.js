export { UserList } from './userList.component';
