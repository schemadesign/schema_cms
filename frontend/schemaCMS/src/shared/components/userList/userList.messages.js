/* eslint-disable max-len */
import { defineMessages } from 'react-intl';

import { ROLES } from '../../../modules/userProfile/userProfile.constants';

const { ADMIN, EDITOR } = ROLES;

export default defineMessages({
  [ADMIN]: {
    id: `shared.components.userList.${ADMIN}`,
    defaultMessage: 'Admin',
  },
  [EDITOR]: {
    id: `shared.components.userList.${EDITOR}`,
    defaultMessage: 'Editor',
  },
});
