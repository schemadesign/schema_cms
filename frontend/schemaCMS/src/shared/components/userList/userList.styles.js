import styled from 'styled-components';
import { Link } from 'react-router-dom';

import { ListItem as ListItemBase } from '../listComponents';
import { media } from '../../../theme/media';

export const UserDetails = styled.div`
  flex: 1 1 auto;
`;

export const UserFullName = styled(Link)`
  color: ${({ theme: { text } }) => text};
  text-decoration: none;
  font-size: 24px;
  font-weight: 600;
  word-break: break-word;
`;

export const ListItem = styled(ListItemBase)`
  min-height: 94px;

  ${media.desktop`
    min-height: 88px;
  `};
`;

export const ListItemContent = styled.div`
  display: flex;
  flex-direction: row;
  align-content: space-between;
`;

export const Role = styled.div`
  color: ${({ theme: { card } }) => card.label};
  font-weight: bold;
`;

export const cardStyles = {
  flexDirection: null,
};

export const buttonIconStyles = {
  height: 30,
  width: 30,
  minHeight: 30,
  padding: 0,
  position: 'absolute',
  top: 10,
  right: 10,
  backgroundColor: 'transparent',
};
