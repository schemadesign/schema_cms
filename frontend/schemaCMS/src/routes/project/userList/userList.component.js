import React, { Fragment, PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { always } from 'ramda';

import { UserList as UserListComponent } from '../../../shared/components/userList';
import {
  BackArrowButton,
  NavigationContainer,
  PlusLink,
  LARGE_BUTTON_SIZE,
} from '../../../shared/components/navigation';
import { ProjectTabs } from '../../../shared/components/projectTabs';
import { USERS } from '../../../shared/components/projectTabs/projectTabs.constants';
import { ContextHeader } from '../../../shared/components/contextHeader';
import messages from './userList.messages';
import browserHistory from '../../../shared/utils/history';
import { renderWhenTrue } from '../../../shared/utils/rendering';
import { LoadingWrapper } from '../../../shared/components/loadingWrapper';
import { getMatchParam, filterMenuOptions } from '../../../shared/utils/helpers';
import reportError from '../../../shared/utils/reportError';
import { MobileMenu } from '../../../shared/components/menu/mobileMenu';
import { getProjectMenuOptions, PROJECT_USERS_ID } from '../project.constants';

export class UserList extends PureComponent {
  static propTypes = {
    userRole: PropTypes.string,
    fetchUsers: PropTypes.func.isRequired,
    removeUser: PropTypes.func.isRequired,
    users: PropTypes.array.isRequired,
    isAdmin: PropTypes.bool.isRequired,
    match: PropTypes.shape({
      params: PropTypes.object.isRequired,
    }).isRequired,
  };

  state = {
    error: null,
    loading: true,
  };

  async componentDidMount() {
    try {
      const projectId = getMatchParam(this.props, 'projectId');

      await this.props.fetchUsers({ projectId });
      this.setState({ loading: false });
    } catch (error) {
      reportError(error);
      this.setState({ loading: false, error });
    }
  }

  getAddUserUrl = () => {
    const projectId = getMatchParam(this.props, 'projectId');
    return `/project/${projectId}/user/add`;
  };

  handleBackClick = () => browserHistory.push(`/project/${getMatchParam(this.props, 'projectId')}`);

  renderCreateUserButton = ({ id, isAdmin }) =>
    renderWhenTrue(always(<PlusLink id={id} to={this.getAddUserUrl()} size={LARGE_BUTTON_SIZE} />))(isAdmin);

  render() {
    const { error, loading } = this.state;
    const { match, isAdmin, users, userRole } = this.props;
    const headerTitle = <FormattedMessage {...messages.headerTitle} />;
    const headerSubtitle = <FormattedMessage {...messages.headerSubtitle} />;
    const projectId = getMatchParam(this.props, 'projectId');
    const menuOptions = getProjectMenuOptions(projectId);

    return (
      <Fragment>
        <MobileMenu
          headerTitle={headerTitle}
          headerSubtitle={headerSubtitle}
          options={filterMenuOptions(menuOptions, userRole)}
          active={PROJECT_USERS_ID}
        />
        <ProjectTabs active={USERS} url={`/project/${match.params.projectId}`} />
        <ContextHeader title={headerTitle} subtitle={headerSubtitle}>
          {this.renderCreateUserButton({ id: 'addUserDesktopBtn', isAdmin })}
        </ContextHeader>
        <LoadingWrapper
          loading={loading}
          noData={!users.length}
          noDataContent={<FormattedMessage {...messages.noUsers} />}
          error={error}
        >
          <UserListComponent users={this.props.users} projectId={getMatchParam(this.props, 'projectId')} />
        </LoadingWrapper>
        <NavigationContainer fixed hideOnDesktop>
          <BackArrowButton onClick={this.handleBackClick}>
            <FormattedMessage {...messages.back} />
          </BackArrowButton>
          {this.renderCreateUserButton({ id: 'addUserBtn', isAdmin })}
        </NavigationContainer>
      </Fragment>
    );
  }
}
