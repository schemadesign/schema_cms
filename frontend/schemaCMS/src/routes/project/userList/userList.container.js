import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { withRouter } from 'react-router-dom';
import { hot } from 'react-hot-loader';
import { injectIntl } from 'react-intl';
import { compose } from 'ramda';
import { promisifyRoutine, bindPromiseCreators } from 'redux-saga-routines';

import { UserList } from './userList.component';
import { ProjectRoutines, selectEditors } from '../../../modules/project';
import { selectIsAdmin, selectUserRole } from '../../../modules/userProfile';

const mapStateToProps = createStructuredSelector({
  userRole: selectUserRole,
  users: selectEditors,
  isAdmin: selectIsAdmin,
});

export const mapDispatchToProps = dispatch =>
  bindPromiseCreators(
    {
      fetchUsers: promisifyRoutine(ProjectRoutines.fetchEditors),
      removeUser: promisifyRoutine(ProjectRoutines.removeEditor),
    },
    dispatch
  );

export default compose(
  hot(module),
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  injectIntl,
  withRouter
)(UserList);
