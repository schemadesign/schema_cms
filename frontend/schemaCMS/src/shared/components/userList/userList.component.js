import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Button, Icons } from 'schemaUI';
import { ifElse, is } from 'ramda';

import { renderWhenTrue } from '../../utils/rendering';
import { Actions, UserDetails, UserFullName, Email, buttonIconStyles, iconStyles, cardStyles } from './userList.styles';

import { ListItem, ListContainer } from '../listComponents';

const { CloseIcon } = Icons;

export class UserList extends PureComponent {
  static propTypes = {
    users: PropTypes.array.isRequired,
    isAdmin: PropTypes.bool.isRequired,
    projectId: PropTypes.string,
    onRemoveUser: PropTypes.func,
  };

  getUrl = ifElse(is(String), projectId => `/project/${projectId}/user/`, () => '/user/');

  renderRemove = renderWhenTrue((_, user) => (
    <Button customStyles={buttonIconStyles} onClick={this.props.onRemoveUser(user)}>
      <CloseIcon customStyles={iconStyles} />
    </Button>
  ));

  renderListItem = ({ user, hasRemoving, isAdmin, index }) => (
    <ListItem key={index} customStyles={cardStyles}>
      <UserDetails>
        <UserFullName to={`${this.getUrl(this.props.projectId)}${user.id}`}>
          {user.firstName} {user.lastName}
        </UserFullName>
        <Email>{user.email}</Email>
      </UserDetails>
      <Actions>{this.renderRemove(hasRemoving && isAdmin, user)}</Actions>
    </ListItem>
  );

  render() {
    const { users, onRemoveUser, isAdmin } = this.props;
    const hasRemoving = !!onRemoveUser;

    return (
      <ListContainer>
        {users.map((user, index) => this.renderListItem({ user, hasRemoving, isAdmin, index }))}
      </ListContainer>
    );
  }
}
