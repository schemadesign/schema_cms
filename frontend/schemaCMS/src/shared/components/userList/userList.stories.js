import React from 'react';
import { storiesOf } from '@storybook/react';

import { UserList } from './userList.component';

const defaultProps = {};

storiesOf('UserList', module).add('Default', () => <UserList {...defaultProps} />);
