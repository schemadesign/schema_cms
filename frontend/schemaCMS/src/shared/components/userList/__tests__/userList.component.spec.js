import React from 'react';
import { shallow } from 'enzyme';

import { UserListComponent as UserList } from '../userList.component';

describe('UserList: Component', () => {
  const defaultProps = {};

  const component = props => <UserList {...defaultProps} {...props} />;

  const render = (props = {}) => shallow(component(props));

  it('should render correctly', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });
});
