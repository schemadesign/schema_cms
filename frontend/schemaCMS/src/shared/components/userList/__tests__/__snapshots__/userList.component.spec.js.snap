// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList: Component should render correctly 1`] = `<styled.section />`;

exports[`UserList: Component should render correctly with users 1`] = `
<styled.section>
  <Styled(Styled(Component))
    customStyles={
      Object {
        "flexDirection": null,
      }
    }
    headerComponent={
      <React.Fragment>
        <ForwardRef(styled.div)>
          <Memo(FormattedMessage)
            defaultMessage="Admin"
            id="shared.components.userList.admin"
          />
        </ForwardRef(styled.div)>
      </React.Fragment>
    }
    key="0"
  >
    <styled.div>
      <styled.div>
        <Styled(Link)
          to="/user/1"
        >
          Alan
           
          Watts
        </Styled(Link)>
      </styled.div>
    </styled.div>
  </Styled(Styled(Component))>
  <Styled(Styled(Component))
    customStyles={
      Object {
        "flexDirection": null,
      }
    }
    headerComponent={
      <React.Fragment>
        <ForwardRef(styled.div)>
          <Memo(FormattedMessage)
            defaultMessage="Editor"
            id="shared.components.userList.editor"
          />
        </ForwardRef(styled.div)>
      </React.Fragment>
    }
    key="1"
  >
    <styled.div>
      <styled.div>
        <Styled(Link)
          to="/user/2"
        >
          Dale
           
          Chihuly
        </Styled(Link)>
      </styled.div>
    </styled.div>
  </Styled(Styled(Component))>
  <Styled(Styled(Component))
    customStyles={
      Object {
        "flexDirection": null,
      }
    }
    headerComponent={
      <React.Fragment>
        <ForwardRef(styled.div)>
          <Memo(FormattedMessage)
            defaultMessage="Admin"
            id="shared.components.userList.admin"
          />
        </ForwardRef(styled.div)>
      </React.Fragment>
    }
    key="2"
  >
    <styled.div>
      <styled.div>
        <Styled(Link)
          to="/user/3"
        >
          Dave
           
          Bowie
        </Styled(Link)>
      </styled.div>
    </styled.div>
  </Styled(Styled(Component))>
</styled.section>
`;
